Add Dependabot auto-merge workflow and migration script

Marenz opened the pull request:
## Summary

- Adds auto-merge workflow for Dependabot PRs using `frequenz-floss/dependabot-auto-approve` action
- Includes migration script to create workflow file and disable CODEOWNERS review requirement via GitHub API
- Splits changes into two commits: reset migration script to template, then add new migration steps

## Changes

**First commit:** Reset `cookiecutter/migrate.py` to template, removing old migration steps

**Second commit:**
- Add `create_dependabot_auto_merge_workflow()` function
- Add `disable_codeowners_review_requirement()` function to update GitHub rulesets
- Include workflow template in cookiecutter
- Regenerate golden test files

Marenz commented:
While testing in the FCR Repo I found it still didn't work and added the fix you see here, after which it finally worked

Marenz commented:
Adds `auto-merge: 'true'` to the Dependabot workflow template.

Migration script fixes:
- Was skipping existing files instead of updating them
- Was matching job name `auto-merge:` instead of parameter `auto-merge: 'true'`
- Only checked `auto-dependabot.yaml`, now searches all workflow files

The script writes the updated workflow file directly - users still need to review, commit with DCO+GPG, and create a PR.

llucax commented on cookiecutter/{{cookiecutter.github_repo_name}}/.github/workflows/auto-dependabot.yaml:
```diff
@@ -0,0 +1,22 @@
+          dependency-type: 'all'
+          auto-merge: 'true'
+          merge-method: 'merge'
+          add-label: 'auto-merged'
```
We are using the scheme `prefix:label`, I would use a prefix here too. So far prefix is one of: `part`, `scope`, `status`, `type`, `priority` and `cmd`. I guess none of those are a good fit for this, maybe we can add `tool` prefix for labels added by tools doing stuff automatically? So something like `tool:auto-merged`?

llucax replied:
Also missing EOL

llucax commented on cookiecutter/{{cookiecutter.github_repo_name}}/.github/workflows/auto-dependabot.yaml:
```diff
@@ -0,0 +1,22 @@
+name: Auto-merge Dependabot PRs
```
Nitpick: Use singular, as this approves and merges PRs one-by-one? It might be more clear what it does when looking at the GitHub "Actions" tab.


```suggestion
name: Auto-merge Dependabot PR
```

Marenz committed:
Reset migration script to template

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,190 @@
+            print(f"Workflow file {workflow_file} already has auto-merge parameter.")
+            return
+        print(f"Workflow file {workflow_file} missing auto-merge parameter.")
+        print("Updating existing workflow file...")
```
This seems to be pretty complex, I would just blindly overwrite with the new version, if you want to be extra nice, just let the user know you replaced the file, but they will also find out when looking at the diff.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,164 @@
+
+def disable_codeowners_review_requirement() -> None:  # noqa: C901
+    """Disable CODEOWNERS review requirement in GitHub repository ruleset."""
+    import json  # pylint: disable=import-outside-toplevel
```
I agree, I would just move to the top.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,190 @@
+    print(f"{action} Dependabot auto-merge workflow at {workflow_file}")
+
+
+def disable_codeowners_review_requirement() -> None:  # noqa: C901
```
This is nice but it will fail for any non-admin user, right? Not sure if it makes a lot of sense to include it, if it fails gracefully and it is already done, I guess it is fine to keep it.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,190 @@
+    except subprocess.CalledProcessError as e:
+        print(f"Failed to get default branch: {e}")
+        print("Skipping CODEOWNERS review requirement update.")
+        print("You may need to manually disable the CODEOWNERS review requirement.")
```
I would remind the user that this requires admin permissions, so they don't pull their hair trying to find the option. Since you have the GitHub URL, you can even provide a URL to go look for the rulesets: `https://github.com/{org}}/{repo}/settings/rules`

llucax replied:
Better use `manual_step()`, so it gets highlighted.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,190 @@
-            continue
+        if not version_branch_ruleset:
+            print("Protect version branches ruleset not found, skipping update.")
+            print("You may need to manually disable the CODEOWNERS review requirement.")
```
Same about `manual_step()`.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,190 @@
+
+    except subprocess.CalledProcessError as e:
+        print(f"Failed to update GitHub ruleset: {e}")
+        print("You may need to manually disable the CODEOWNERS review requirement.")
```
`manual_step()`

llucax commented on uv.lock:
I guess this one is accidental, and if it is not, I would add any `uv` related changes in a separate commit.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,164 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Auto-merge Dependabot PR
+        uses: frequenz-floss/dependabot-auto-approve@3cad5f42e79296505473325ac6636be897c8b8a1 # v1.3.2  # noqa: E501
```
This

llucax commented on cookiecutter/migrate.py:
```diff
@@ -22,203 +22,164 @@
+      - name: Auto-merge Dependabot PR
+        uses: frequenz-floss/dependabot-auto-approve@3cad5f42e79296505473325ac6636be897c8b8a1 # v1.3.2  # noqa: E501
+        with:
+          github-token: {% raw %}${{ secrets.GITHUB_TOKEN }}{% endraw %}
```
And this.

Marenz committed:
Add Dependabot auto-merge workflow and CODEOWNERS bypass

- Add auto-dependabot.yaml workflow to auto-merge Dependabot PRs
- Implement migration script functions to create workflow files
- Add GitHub API integration to disable CODEOWNERS review requirement
- Update all cookiecutter templates and golden test files

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Add release notes for Dependabot auto-merge feature

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz commented:
Updated the branch with some refactoring and cleanup:

- Refactored the migration script by extracting helper functions (`get_default_branch()`, `find_version_branch_ruleset()`, `update_ruleset()`) for better code organization and error handling
- Improved error messages with proper `manual_step()` calls that provide direct GitHub settings URLs when API operations fail
- Standardized the Dependabot workflow formatting across all templates and golden test files (consistent naming, label changed to `tool:auto-merged`, YAML formatting)
- Fixed the release notes - they were accidentally removed in the previous iteration, now properly documenting the new auto-merge feature
- Removed `uv.lock` that shouldn't have been committed

llucax commented on cookiecutter/{{cookiecutter.github_repo_name}}/.github/workflows/auto-dependabot.yaml:
```diff
@@ -0,0 +1,22 @@
+      - name: Auto-merge Dependabot PR
+        uses: frequenz-floss/dependabot-auto-approve@3cad5f42e79296505473325ac6636be897c8b8a1 # v1.3.2
+        with:
+          github-token: {% raw %}${{ secrets.GITHUB_TOKEN }}{% endraw %}
```
Nitpick: To avoid any other issues with template interpretation, other workflows actually just wrap the whole file between `{% raw %}` and `{% endraw %}`, maybe it would be better to do it in this file too, as it would be rare that we want to use any other template features in it.

llucax commented on cookiecutter/migrate.py:
```diff
@@ -21,204 +21,228 @@
+        )
+        return True
+    except subprocess.CalledProcessError:
+        return False
```
Nitpick: Print the exception for more context of why it failed?

llucax reviewed (dismissed):
2 small comments, feel free to force-merge if you want to fix them.
